### DISCUSSION TITLE
More 'query' examples

### DISCUSSION THREAD

**Discussion by sombriks** [Level/community#124] [General]
As far as i understand, level is key/value based and therefore there is no query api except for some `RangeOptions` which we can use with `.values()`  and `.iterator()`

But would be nice to find more examples, maybe even going beyond those options.

**Comment by juliangruber** [comment]
Can you think of a query that you would like to see?

**Reply by sombriks** [reply]
i was thinking about find a substring on my values. 

for example, if i save my data like this:

```javascript
export const updateTodoService = async ({ key, message, done }) => {
  const todo = { key, message, done }
  await db.put(key, todo)
  return todo
}
```

i was looking for a way to get values based on, for example, todo's containing _hello_ in message.

so far i managed to expose what `ValueIteratorOptions` has exposed: 

```javascript
export const listTodoService = async (query = {}) => {
  const { limit = 500, reverse = true } = query
  return await db.values({ ...query, limit: parseInt(limit), reverse: "true" == `${reverse}` }).all()
}
```

but i am failing to find docs or examples filtering based on my values.

any guidance is much welcome!

the sample code i am tweaking is [here](https://github.com/sombriks/node-simple).

**Reply by vweevers** [reply]
Compare it to a relational database (RDBMS). You can query such a database, but if you don't have any indexes, it'll do a "full table scan" - which means to read every table row in order to find matching data. It's the same here: to filter values (aka rows) you must either read all of them, or create an index. For the latter, you can roll your own (essentially turning values into keys so that you can perform range queries) or look through [awesome](https://github.com/Level/awesome#processing) modules (might be dated).

**Reply by sombriks** [reply]
Hey, thanks for the guidance! Indeed i need to rethink how am i supposed to deal with the keys and realize how level works if i want to get the best of it!
